File: mobile-android/core-logic/src/main/kotlin/com/mesadaapp/mobile/core/ResolvedorConflito.kt
package com.mesadaapp.mobile.core

/**
 * "Conflitos (ex.: tarefa alterada na Web enquanto o app estava offline)
 * resolvidos por timestamp — a mudança mais recente prevalece, com o Master
 * notificado se um conflito de valor ocorrer." (docs/adendo-mobile.md)
 */
enum class VencedorConflito { LOCAL, REMOTO }

data class ResultadoConflito(
    val vencedor: VencedorConflito,
    /** true quando a versão local foi descartada — o Master precisa ser avisado. */
    val requerNotificacaoAoMaster: Boolean,
)

object ResolvedorConflito {

    /**
     * @param timestampLocal momento em que o filho marcou a execução no aparelho, offline.
     * @param timestampRemoto momento da última alteração dessa mesma execução no backend.
     *
     * Empate (timestamps iguais) é tratado como vitória do remoto: não há
     * como provar que a marcação local é estritamente mais recente, então o
     * lado servidor — que já é a fonte de verdade para o Master — prevalece,
     * e o filho é avisado de que precisa conferir/refazer.
     */
    fun resolver(timestampLocal: Long, timestampRemoto: Long): ResultadoConflito =
        if (timestampLocal > timestampRemoto) {
            ResultadoConflito(vencedor = VencedorConflito.LOCAL, requerNotificacaoAoMaster = false)
        } else {
            ResultadoConflito(vencedor = VencedorConflito.REMOTO, requerNotificacaoAoMaster = true)
        }
}

File: mobile-ios/MesadaCore/Sources/MesadaCore/ResolvedorConflito.swift
import Foundation

/// "Conflitos (ex.: tarefa alterada na Web enquanto o app estava offline)
/// resolvidos por timestamp — a mudança mais recente prevalece, com o
/// Master notificado se um conflito de valor ocorrer."
/// (docs/adendo-mobile.md) — mesma regra de
/// com.mesadaapp.mobile.core.ResolvedorConflito (Android).
public enum VencedorConflito: Sendable {
    case local
    case remoto
}

public struct ResultadoConflito: Sendable {
    public let vencedor: VencedorConflito
    /// true quando a versão local foi descartada — o Master precisa ser avisado.
    public let requerNotificacaoAoMaster: Bool
}

public enum ResolvedorConflito {

    /// Empate é tratado como vitória do remoto: o servidor é a fonte de
    /// verdade do Master, então em caso de dúvida ele prevalece e o filho é
    /// avisado para conferir/refazer.
    public static func resolver(timestampLocal: Date, timestampRemoto: Date) -> ResultadoConflito {
        if timestampLocal > timestampRemoto {
            return ResultadoConflito(vencedor: .local, requerNotificacaoAoMaster: false)
        } else {
            return ResultadoConflito(vencedor: .remoto, requerNotificacaoAoMaster: true)
        }
    }
}
